### ISSUE TITLE
Fetching metadata on subscribe causing high memory on client and server

### ISSUE THREAD

**Issue by cruickshankpg** [liftbridge-io/go-liftbridge#100]
When subscribing to a stream, if the stream is not in the `metadataCache` the cache is completely updated resulting in the `FetchMedata` RPC being called. This can happen if the stream is new or does not exist.

https://github.com/liftbridge-io/go-liftbridge/blob/c842e2f19749a5aa5a7664b1a6d4a366034d8f92/v2/client.go#L1518

We only create a stream when a message is published to save on unnecessary creates but this means that if multiple subscribers attempt to subscribe before the publisher then a lot of `FetchMetadata` RPCs are made. When there are 1000s (I had 3000 when I hit this) of streams in the liftbridge cluster the metadata gets very big and marshalling it so frequently caused one of my liftbridge servers to become unresponsive and all the memory on my client to be used up.

Our liftbridge client service has liftbridge client connections to multiple liftbridge clusters so even storing the full metadata for each cluster is more memory than we would like. Keeping track of the brokers for a cluster is obviously necessary but do we need all the streams? Could individual stream partitions be fetched into the cache on demand?

**Comment by tylertreat** [comment]
>  Keeping track of the brokers for a cluster is obviously necessary but do we need all the streams? Could individual stream partitions be fetched into the cache on demand?

Yes, this is an area for improvement I've had in mind. The client should only fetch the streams it needs. Also, the `FetchMetadata` RPC [already supports this](https://github.com/liftbridge-io/liftbridge-api/blob/8ceb6d0858acc507a5694d5f1a899f714cbfaeb8/api.proto#L116). It just defaults to fetching everything if no streams are specified, so it should be a fairly simple change.
